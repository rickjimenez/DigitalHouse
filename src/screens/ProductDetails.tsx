import React, {FC} from 'react';
import {
  Text,
  View,
  StyleSheet,
  Image,
  SafeAreaView,
  ScrollView,
} from 'react-native';
import {ProductDetailsProps} from '../types/navigation';
import {useQueryClient} from '@tanstack/react-query';
import {GetProductsResponse} from '../types/api';
import Button from '../components/Button';
import {useNavigation} from '@react-navigation/native';
import format from 'date-fns/format';
import es from 'date-fns/locale/es';

/**
 * ProductDetails Component
 *
 * @returns {React.ReactElement} ProductDetails.
 */
const ProductDetails: FC<ProductDetailsProps> = ({route}): JSX.Element => {
  const styles = useStyles();
  const nav = useNavigation<ProductDetailsProps['navigation']>();
  const queryClient = useQueryClient();
  const data: GetProductsResponse[] | undefined = queryClient.getQueryData([
    'products',
  ]);

  const [item] = data?.filter(e => e.id === route.params.productId);
  return (
    <View style={styles.container}>
      <SafeAreaView style={styles.customHeader}>
        <Text style={styles.title}>{item.product}</Text>
      </SafeAreaView>

      <ScrollView>
        <View style={styles.imageContainer}>
          <Image style={styles.image} source={{uri: item.image}} />
        </View>

        <View style={styles.productInfoContainer}>
          <Text style={styles.label}>Detalles del producto:</Text>
          <Text style={styles.date}>
            Comprado el{' '}
            {format(new Date(item.createdAt), "dd 'de' MMMM yyyy", {
              locale: es,
            })}
          </Text>
          <Text style={styles.label}>{`Con esta compra ${
            item.is_redemption ? 'canjeaste' : 'acumulaste'
          }:`}</Text>
          <Text style={styles.points}>{item.points} Puntos</Text>
          <Button
            label="Aceptar"
            onPress={() => nav.goBack()}
            testID="back-btn"
          />
        </View>
      </ScrollView>
    </View>
  );
};

const useStyles = () =>
  StyleSheet.create({
    container: {
      flex: 1,
    },
    customHeader: {
      backgroundColor: '#CFD6FF',
      height: 150,
      justifyContent: 'flex-end',
    },
    title: {
      fontSize: 24,
      fontWeight: '800',
      marginBottom: 20,
      marginLeft: 20,
    },
    productInfoContainer: {
      paddingHorizontal: 20,
    },
    label: {
      fontSize: 14,
      fontWeight: '800',
      color: '#9B9898',
    },
    date: {
      fontSize: 16,
      fontWeight: '800',
      marginVertical: 20,
    },
    points: {
      fontSize: 24,
      fontWeight: '800',
      marginVertical: 40,
    },
    image: {
      width: '100%',
      height: 350,
      borderRadius: 10,
    },
    imageContainer: {
      shadowColor: '#000',
      shadowOffset: {
        width: 0,
        height: 2,
      },
      shadowOpacity: 0.25,
      shadowRadius: 3.84,

      elevation: 5,
      padding: 20,
    },
  });

export default ProductDetails;
