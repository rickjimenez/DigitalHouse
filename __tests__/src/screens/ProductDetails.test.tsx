import 'react-native';
import {fireEvent, render, screen} from '@testing-library/react-native';
import React from 'react';
import ProductDetails from '../../../src/screens/ProductDetails';

const data = [
  {
    id: '1',
    product: 'product 1',
    is_redemption: false,
    createdAt: '2022-12-09T06:34:25.607Z',
    image: 'image',
    points: 100,
  },
  {
    id: '2',
    product: 'product 2',
    is_redemption: true,
    createdAt: '2022-12-09T06:34:25.607Z',
    image: 'image',
    points: 30,
  },
];

jest.mock('@tanstack/react-query', () => ({
  ...jest.requireActual('@tanstack/react-query'),
  useQueryClient: () => ({
    setQueryData: jest.fn(),
    getQueryData: jest.fn().mockReturnValueOnce(data).mockReturnValueOnce(data),
  }),
}));

const mockedNavigate = jest.fn();

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({navigate: mockedNavigate, goBack: mockedNavigate}),
}));

describe('test TotalPoints Comp', () => {
  test('renders correctly', async () => {
    render(<ProductDetails route={{params: {productId: '1'}}} />);

    const comp = screen.toJSON();
    expect(comp).toMatchSnapshot();
  });

  test('test go back', async () => {
    render(<ProductDetails route={{params: {productId: '2'}}} />);
    const btn = screen.getByTestId('back-btn');
    fireEvent.press(btn);
    expect(mockedNavigate).toHaveBeenCalledTimes(1);
  });
});
