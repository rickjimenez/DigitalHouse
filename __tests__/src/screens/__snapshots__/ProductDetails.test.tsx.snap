// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test TotalPoints Comp renders correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <RCTSafeAreaView
    style={
      {
        "backgroundColor": "#CFD6FF",
        "height": 150,
        "justifyContent": "flex-end",
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 24,
          "fontWeight": "800",
          "marginBottom": 20,
          "marginLeft": 20,
        }
      }
    >
      product 1
    </Text>
  </RCTSafeAreaView>
  <RCTScrollView>
    <View>
      <View
        style={
          {
            "elevation": 5,
            "padding": 20,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.25,
            "shadowRadius": 3.84,
          }
        }
      >
        <Image
          source={
            {
              "uri": "image",
            }
          }
          style={
            {
              "borderRadius": 10,
              "height": 350,
              "width": "100%",
            }
          }
        />
      </View>
      <View
        style={
          {
            "paddingHorizontal": 20,
          }
        }
      >
        <Text
          style={
            {
              "color": "#9B9898",
              "fontSize": 14,
              "fontWeight": "800",
            }
          }
        >
          Detalles del producto:
        </Text>
        <Text
          style={
            {
              "fontSize": 16,
              "fontWeight": "800",
              "marginVertical": 20,
            }
          }
        >
          Comprado el
           
          09 de diciembre 2022
        </Text>
        <Text
          style={
            {
              "color": "#9B9898",
              "fontSize": 14,
              "fontWeight": "800",
            }
          }
        >
          Con esta compra acumulaste:
        </Text>
        <Text
          style={
            {
              "fontSize": 24,
              "fontWeight": "800",
              "marginVertical": 40,
            }
          }
        >
          100
           Puntos
        </Text>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#334FFA",
              "borderRadius": 10,
              "height": 50,
              "justifyContent": "center",
              "opacity": 1,
            }
          }
          testID="back-btn"
        >
          <Text
            style={
              {
                "color": "white",
                "fontSize": 14,
                "fontWeight": "800",
              }
            }
          >
            Aceptar
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
